File: backend/api/routes/boq.py
"""POST /api/boq — wraps backend/engine/boq_calculator.py directly."""

from fastapi import APIRouter, HTTPException

from backend.api.schemas import BoQLineItemResponse, BoQRequest, BoQResponse
from backend.engine.boq_calculator import ZoneInput, calculate_boq

router = APIRouter()


def _build_zone_input(zone) -> ZoneInput:
    if zone.unit == "Rft":
        return ZoneInput(
            zone_id=zone.zone_id,
            zone_name=zone.zone_name,
            gross_area=zone.running_feet,
            material_id=zone.material_id,
            openings=0.0,
        )
    return ZoneInput(
        zone_id=zone.zone_id,
        zone_name=zone.zone_name,
        gross_area=zone.gross_area_sqft,
        material_id=zone.material_id,
        openings=zone.deductions_sqft,
    )


def _build_rate_overrides(request: BoQRequest) -> dict:
    """Translates per-zone overrides (keyed by zoneId, as the frontend sends them)
    into the engine's per-material overrides (keyed by materialId)."""
    overrides: dict = {}
    for zone in request.zones:
        override = request.rate_overrides.get(zone.zone_id)
        if override is None:
            continue
        entry = overrides.setdefault(zone.material_id, {})
        if override.material_rate_inr is not None:
            entry["material_rate"] = override.material_rate_inr
        if override.labor_rate_inr is not None:
            entry["labor_rate"] = override.labor_rate_inr
    return overrides


@router.post("/boq", response_model=BoQResponse, response_model_by_alias=True)
def compute_boq(request: BoQRequest) -> BoQResponse:
    zone_inputs = [_build_zone_input(z) for z in request.zones]
    rate_overrides = _build_rate_overrides(request)

    try:
        summary = calculate_boq(
            zone_inputs,
            rate_overrides=rate_overrides,
            contingency_percentage=request.contingency_pct / 100.0,
        )
    except KeyError as e:
        raise HTTPException(status_code=400, detail=str(e))

    return BoQResponse(
        items=[
            BoQLineItemResponse(
                zone_id=item.zone_id,
                zone_name=item.zone_name,
                material_id=item.material_id,
                material_name=item.material_name,
                unit=item.unit,
                net_area=item.net_workable_area,
                wastage_pct=item.wastage_percentage,
                gross_material_area=item.gross_material_area,
                unit_material_rate_inr=item.unit_material_rate_inr,
                unit_labor_rate_inr=item.unit_labor_rate_inr,
                material_cost_inr=item.material_cost_inr,
                labor_cost_inr=item.labor_cost_inr,
                line_total_inr=item.line_total_inr,
            )
            for item in summary.items
        ],
        total_material_cost_inr=summary.total_material_cost_inr,
        total_labor_cost_inr=summary.total_labor_cost_inr,
        subtotal_inr=summary.subtotal_inr,
        contingency_pct=summary.contingency_percentage,
        contingency_amount_inr=summary.contingency_amount_inr,
        grand_total_inr=summary.grand_total_inr,
    )

File: frontend/src/lib/boq.ts
// Port of backend/engine/boq_calculator.py's calculate_line_item / calculate_boq.
// Key rule (verified against the Python source): material cost is charged on the
// wastage-inflated GROSS area, labor cost is charged on the NET (as-installed) area.
import type { BoQGroup, BoQLineItem, BoQSummary, Material, Zone } from "@/types";

export const LABEL_PLURAL: Record<string, string> = {
  wall: "Walls",
  pillar: "Pillars",
  balcony_railing: "Balcony Railings",
  roof_parapet: "Roof Parapets",
  window: "Windows",
  door: "Doors",
};

const round2 = (n: number) => Math.round((n + Number.EPSILON) * 100) / 100;

export function calculateLineItem(
  zone: Zone,
  material: Material,
  rateOverride?: { materialRateInr?: number; laborRateInr?: number }
): BoQLineItem {
  const isLinear = material.unit === "Rft";
  const netArea = isLinear ? zone.runningFeet : zone.netAreaSqft;

  const wastageAllowance = round2(netArea * material.wastageFactor);
  const grossMaterialArea = round2(netArea + wastageAllowance);

  const matRate = rateOverride?.materialRateInr ?? material.materialRateInr;
  const labRate = rateOverride?.laborRateInr ?? material.laborRateInr;

  const materialCostInr = round2(grossMaterialArea * matRate);
  const laborCostInr = round2(netArea * labRate);
  const lineTotalInr = round2(materialCostInr + laborCostInr);

  return {
    zoneId: zone.id,
    zoneName: zone.displayName,
    materialId: material.id,
    materialName: material.shortName,
    unit: material.unit,
    netArea,
    wastagePct: round2(material.wastageFactor * 100),
    grossMaterialArea,
    unitMaterialRateInr: matRate,
    unitLaborRateInr: labRate,
    materialCostInr,
    laborCostInr,
    lineTotalInr,
  };
}

export function calculateBoQ(
  items: BoQLineItem[],
  contingencyPct = 0.05
): BoQSummary {
  const totalMaterialCostInr = round2(items.reduce((s, i) => s + i.materialCostInr, 0));
  const totalLaborCostInr = round2(items.reduce((s, i) => s + i.laborCostInr, 0));
  const subtotalInr = round2(totalMaterialCostInr + totalLaborCostInr);
  const contingencyAmountInr = round2(subtotalInr * contingencyPct);
  const grandTotalInr = round2(subtotalInr + contingencyAmountInr);

  return {
    items,
    totalMaterialCostInr,
    totalLaborCostInr,
    subtotalInr,
    contingencyPct: round2(contingencyPct * 100),
    contingencyAmountInr,
    grandTotalInr,
  };
}

/** Collapses line items that share the same zone label + material into one display row. */
export function groupBoQItems(items: BoQLineItem[], zones: Zone[]): BoQGroup[] {
  const zoneById = new Map(zones.map((z) => [z.id, z]));
  const groups = new Map<string, BoQGroup>();

  for (const item of items) {
    const zone = zoneById.get(item.zoneId);
    const label = zone?.label ?? "zone";
    const key = `${label}::${item.materialId}`;
    const existing = groups.get(key);

    if (existing) {
      existing.zoneIds.push(item.zoneId);
      existing.zoneCount += 1;
      existing.netArea = round2(existing.netArea + item.netArea);
      existing.materialCostInr = round2(existing.materialCostInr + item.materialCostInr);
      existing.laborCostInr = round2(existing.laborCostInr + item.laborCostInr);
      existing.lineTotalInr = round2(existing.lineTotalInr + item.lineTotalInr);
      existing.displayName = `${LABEL_PLURAL[label] ?? label} ×${existing.zoneCount}`;
    } else {
      groups.set(key, {
        groupKey: key,
        zoneIds: [item.zoneId],
        zoneCount: 1,
        displayName: item.zoneName,
        materialId: item.materialId,
        materialName: item.materialName,
        unit: item.unit,
        netArea: item.netArea,
        wastagePct: item.wastagePct,
        unitMaterialRateInr: item.unitMaterialRateInr,
        unitLaborRateInr: item.unitLaborRateInr,
        materialCostInr: item.materialCostInr,
        laborCostInr: item.laborCostInr,
        lineTotalInr: item.lineTotalInr,
      });
    }
  }

  return Array.from(groups.values()).sort((a, b) => b.lineTotalInr - a.lineTotalInr);
}

export function formatINR(amount: number): string {
  return new Intl.NumberFormat("en-IN", {
    style: "currency",
    currency: "INR",
    maximumFractionDigits: 0,
  }).format(amount);
}
